libstb/secvar/backend/edk2-compat-util.c

```diff
@@ -0,0 +1,601 @@
+// SPDX-License-Identifier: Apache-2.0
+/* Copyright 2019 IBM Corp. */
+#ifndef pr_fmt
+#define pr_fmt(fmt) "EDK2_COMPAT: " fmt
+#endif
+
+#include <opal.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <ccan/endian/endian.h>
+#include <mbedtls/error.h>
+#include <device.h>
+#include "libstb/crypto/pkcs7/pkcs7.h"
+#include "edk2.h"
+#include "../secvar.h"
+#include "../secvar_devtree.h"
+#include "edk2-compat-util.h"
+
+char *utf8_to_ucs2(const char *key, size_t keylen)
+{
+	int i;
+	char *str;
+
+	str = zalloc(keylen * 2);
+	if (!str)
+		return NULL;
+
+	for (i = 0; i < keylen*2; key++) {
+		str[i++] = *key;
+		str[i++] = '\0';
+	}
+
+	return str;
+}
+
+bool key_equals(const char *key1, const char *key2)
+{
+	if (!strncmp(key1, key2, strlen(key1)))
```

Comment:
Using `strnlen()` with a length of a call to `strlen()` defeats the purpose of the bound limit of `strnlen()`. If `key1` is not properly null-terminated, this will still overrun. Should we maybe just make `key_equals()` an inline function or macro to `strnlen()` with a max expected key length?

```C
#define EDK2_MAX_KEY_LEN 5 // semi-arbitrary, but we have no keys longer than this, null-included
#define key_equals(a,b) !strnlen(a,b,EDK2_MAX_KEY_LEN)
```

Reply:
This is strncmp and not strnlen() . Is your comment valid for strncmp as well ?

Reply:
Whoops, updated comment. Meant `strncmp()` in the example